library/include/modules/Maps.h

```diff
@@ -180,6 +180,49 @@ namespace DFHack
         return (p.x & ~15) == 0 && (p.y & ~15) == 0;
     }
 
+    /**
+     * Utility class representing a cuboid of df::coord
+     * \ingroup grp_maps
+     */
+    class cuboid
+    {
+        public:
+        int16_t x_min = -1;
+        int16_t x_max = -1;
+        int16_t y_min = -1;
+        int16_t y_max = -1;
+        int16_t z_min = -1;
+        int16_t z_max = -1;
+
+        // Default constructor
+        DFHACK_EXPORT cuboid() {}
+
+        // Construct from two corners
+        DFHACK_EXPORT cuboid(int16_t x1, int16_t y1, int16_t z1, int16_t x2, int16_t y2, int16_t z2);
+        DFHACK_EXPORT cuboid(const df::coord &p1, const df::coord &p2) { cuboid(p1.x, p1.y, p1.z, p2.x, p2.y, p2.z); }
+
+        // Construct as single tile
+        DFHACK_EXPORT cuboid(int16_t x, int16_t y, int16_t z);
+        DFHACK_EXPORT cuboid(const df::coord &p) { cuboid(p.x, p.y, p.z); }
+
+        // Valid cuboid? True if all max >= min >= 0
+        DFHACK_EXPORT bool isValid() const;
+
+        // Clear cuboid dimensions, making it invalid
+        DFHACK_EXPORT void clear() { x_min = x_max = y_min = y_max = z_min = z_max = -1; }
+
+        // Expand cuboid to include point. Return true if bounds changed
+        DFHACK_EXPORT bool addPos(int16_t x, int16_t y, int16_t z);
+        DFHACK_EXPORT bool addPos(const df::coord &pos) { return addPos(pos.x, pos.y, pos.z); }
+
+        // Return true if point inside cuboid. Make sure cuboid is valid first!
+        DFHACK_EXPORT bool containsPos(int16_t x, int16_t y, int16_t z) const;
+        DFHACK_EXPORT bool containsPos(const df::coord &pos) const { return containsPos(pos.x, pos.y, pos.z); }
+
+        /// Iterate over every point in the cuboid from N-S, W-E, then top-down. Doesn't guarantee valid map tile!
+        /// If z_first iterate top-down, N-S, then W-E.
+        DFHACK_EXPORT void forCoord(std::function<void(df::coord)> fn, bool z_first = false);
```

Comment:
how hard would it be to provide a range interface? I added one for `Units::citizensRange`, and I like how it allows you to break out of the for loop early if you need to. `forCoord` would require that you iterate through all points (or you could change the signature of the callback to return a boolean that would indicate whether to exit early). I guess either option would be fine.

Reply:
Not quite sure how that works. `Units::citizensRange` filters a vector, but we're generating values from nowhere so I need some kind of example.

Reply:
I couldn't find a good example, and buffering the whole sequence in a vector just so you can iterate through them seems like a waste. The second option might be better -- return a boolean from the callback to determine whether you should break out of the for loop early.

Reply:
Tried `break`ing out of the nested for loop. Took me a while to realize the problem with that.